Sync fails on table with column type `YEAR` and not possible to query table

**Describe the bug**
If a MySQL/MariaDB table has a column type `YEAR`, then sync of the table fails and it's not possible to browse/use table.

**To Reproduce**
1. In MySQL/MariaDB create a table with problematic column:
```
DROP TABLE IF EXISTS `x_formatting`;
CREATE TABLE `x_formatting` (
  `ID` bigint(20) UNSIGNED NOT NULL,
  `order_year` year DEFAULT NULL,
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
INSERT INTO `x_formatting` (`ID`, `order_year`) VALUES (1, 2019),(2, 2020);
```
2. Admin > Databases > mysql-db > Sync schema - the log will show sync error (see trace below)
3. Simple question > mysql-db > "x_formatting" errors with `Cannot read LocalDate using a Types.SMALLINT field`

<details><summary>Sync stacktrace</summary>

```
09-03 15:55:44 ERROR sync.util :: Error fingerprinting Table 125 'x_formatting'
clojure.lang.ExceptionInfo: Error reducing result rows {:type :qp}
	at metabase.query_processor.context.default$default_reducef$fn__37927.invoke(default.clj:61)
	at metabase.query_processor.context.default$default_reducef.invokeStatic(default.clj:58)
	at metabase.query_processor.context.default$default_reducef.invoke(default.clj:49)
	at metabase.query_processor.context$reducef.invokeStatic(context.clj:69)
	at metabase.query_processor.context$reducef.invoke(context.clj:62)
	at metabase.query_processor.context.default$default_runf$respond_STAR___37931.invoke(default.clj:70)
	at metabase.driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:392)
	at metabase.driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)
	at metabase.driver.sql_jdbc$fn__73834.invokeStatic(sql_jdbc.clj:49)
	at metabase.driver.sql_jdbc$fn__73834.invoke(sql_jdbc.clj:47)
	at clojure.lang.MultiFn.invoke(MultiFn.java:244)
	at metabase.query_processor.context$executef.invokeStatic(context.clj:59)
	at metabase.query_processor.context$executef.invoke(context.clj:48)
	at metabase.query_processor.context.default$default_runf.invokeStatic(default.clj:69)
	at metabase.query_processor.context.default$default_runf.invoke(default.clj:67)
	at metabase.query_processor.context$runf.invokeStatic(context.clj:45)
	at metabase.query_processor.context$runf.invoke(context.clj:39)
	at metabase.query_processor.reducible$pivot.invokeStatic(reducible.clj:34)
	at metabase.query_processor.reducible$pivot.invoke(reducible.clj:31)
	at metabase.query_processor.middleware.mbql_to_native$mbql__GT_native$fn__45532.invoke(mbql_to_native.clj:26)
	at metabase.query_processor.middleware.check_features$check_features$fn__44834.invoke(check_features.clj:42)
	at metabase.query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__45697.invoke(optimize_datetime_filters.clj:133)
	at metabase.query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47225.invoke(wrap_value_literals.clj:137)
	at metabase.query_processor.middleware.annotate$add_column_info$fn__43455.invoke(annotate.clj:574)
	at metabase.query_processor.middleware.permissions$check_query_permissions$fn__44709.invoke(permissions.clj:64)
	at metabase.query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__46215.invoke(pre_alias_aggregations.clj:40)
	at metabase.query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__44907.invoke(cumulative_aggregations.clj:61)
	at metabase.query_processor.middleware.resolve_joins$resolve_joins$fn__46747.invoke(resolve_joins.clj:183)
	at metabase.query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39185.invoke(add_implicit_joins.clj:245)
	at metabase.query_processor.middleware.limit$limit$fn__45518.invoke(limit.clj:38)
	at metabase.query_processor.middleware.format_rows$format_rows$fn__45499.invoke(format_rows.clj:81)
	at metabase.query_processor.middleware.desugar$desugar$fn__44973.invoke(desugar.clj:22)
	at metabase.query_processor.middleware.binning$update_binning_strategy$fn__43999.invoke(binning.clj:229)
	at metabase.query_processor.middleware.resolve_fields$resolve_fields$fn__44515.invoke(resolve_fields.clj:24)
	at metabase.query_processor.middleware.add_dimension_projections$add_remapping$fn__38734.invoke(add_dimension_projections.clj:272)
	at metabase.query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__38941.invoke(add_implicit_clauses.clj:141)
	at metabase.query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39334.invoke(add_source_metadata.clj:105)
	at metabase.query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__46412.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)
	at metabase.query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__43640.invoke(auto_bucket_datetimes.clj:125)
	at metabase.query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44562.invoke(resolve_source_table.clj:46)
	at metabase.query_processor.middleware.parameters$substitute_parameters$fn__46197.invoke(parameters.clj:114)
	at metabase.query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__44614.invoke(resolve_referenced.clj:80)
	at metabase.query_processor.middleware.expand_macros$expand_macros$fn__45229.invoke(expand_macros.clj:158)
	at metabase.query_processor.middleware.add_timezone_info$add_timezone_info$fn__39365.invoke(add_timezone_info.clj:15)
	at metabase.query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__47109.invoke(splice_params_in_response.clj:32)
	at metabase.query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46423$fn__46427.invoke(resolve_database_and_driver.clj:33)
	at metabase.driver$do_with_driver.invokeStatic(driver.clj:61)
	at metabase.driver$do_with_driver.invoke(driver.clj:57)
	at metabase.query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46423.invoke(resolve_database_and_driver.clj:27)
	at metabase.query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45447.invoke(fetch_source_query.clj:267)
	at metabase.query_processor.middleware.store$initialize_store$fn__47118$fn__47119.invoke(store.clj:11)
	at metabase.query_processor.store$do_with_store.invokeStatic(store.clj:44)
	at metabase.query_processor.store$do_with_store.invoke(store.clj:40)
	at metabase.query_processor.middleware.store$initialize_store$fn__47118.invoke(store.clj:10)
	at metabase.query_processor.middleware.cache$maybe_return_cached_results$fn__44491.invoke(cache.clj:209)
	at metabase.query_processor.middleware.validate$validate_query$fn__47127.invoke(validate.clj:10)
	at metabase.query_processor.middleware.normalize_query$normalize$fn__45545.invoke(normalize_query.clj:22)
	at metabase.query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39203.invoke(add_rows_truncated.clj:36)
	at metabase.query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__47094.invoke(results_metadata.clj:143)
	at metabase.query_processor.reducible$async_qp$qp_STAR___38013$thunk__38014.invoke(reducible.clj:101)
	at metabase.query_processor.reducible$async_qp$qp_STAR___38013.invoke(reducible.clj:107)
	at metabase.query_processor.reducible$async_qp$qp_STAR___38013.invoke(reducible.clj:94)
	at metabase.query_processor.reducible$sync_qp$qp_STAR___38022$fn__38023.invoke(reducible.clj:129)
	at metabase.query_processor.reducible$sync_qp$qp_STAR___38022.invoke(reducible.clj:128)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.query_processor$process_query.invokeStatic(query_processor.clj:122)
	at metabase.query_processor$process_query.doInvoke(query_processor.clj:116)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at clojure.lang.Var.invoke(Var.java:384)
	at metabase.db.metadata_queries$fn__52337$table_rows_sample__52342$fn__52343.invoke(metadata_queries.clj:96)
	at metabase.db.metadata_queries$fn__52337$table_rows_sample__52342.invoke(metadata_queries.clj:92)
	at metabase.sync.analyze.fingerprint$fn__52403$fingerprint_table_BANG___52408$fn__52409.invoke(fingerprint.clj:63)
	at metabase.sync.analyze.fingerprint$fn__52403$fingerprint_table_BANG___52408.invoke(fingerprint.clj:43)
	at metabase.sync.analyze.fingerprint$fn__52576$fingerprint_fields_BANG___52581$fn__52582$fn__52583.invoke(fingerprint.clj:166)
	at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:149)
	at metabase.sync.util$do_with_error_handling.invoke(util.clj:142)
	at metabase.sync.analyze.fingerprint$fn__52576$fingerprint_fields_BANG___52581$fn__52582.invoke(fingerprint.clj:164)
	at metabase.sync.analyze.fingerprint$fn__52576$fingerprint_fields_BANG___52581.invoke(fingerprint.clj:160)
	at metabase.sync.analyze.fingerprint$fn__52604$fingerprint_fields_for_db_BANG___52609$fn__52610$fn__52611$iter__52612__52616$fn__52617$fn__52618.invoke(fingerprint.clj:184)
	at metabase.sync.analyze.fingerprint$fn__52604$fingerprint_fields_for_db_BANG___52609$fn__52610$fn__52611$iter__52612__52616$fn__52617.invoke(fingerprint.clj:181)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.Cons.next(Cons.java:39)
	at clojure.lang.RT.boundedLength(RT.java:1792)
	at clojure.lang.RestFn.applyTo(RestFn.java:130)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.sync.analyze.fingerprint$fn__52604$fingerprint_fields_for_db_BANG___52609$fn__52610$fn__52611.invoke(fingerprint.clj:181)
	at metabase.query_processor.store$do_with_store.invokeStatic(store.clj:46)
	at metabase.query_processor.store$do_with_store.invoke(store.clj:40)
	at metabase.sync.analyze.fingerprint$fn__52604$fingerprint_fields_for_db_BANG___52609$fn__52610.invoke(fingerprint.clj:178)
	at metabase.sync.analyze.fingerprint$fn__52604$fingerprint_fields_for_db_BANG___52609.invoke(fingerprint.clj:172)
	at metabase.sync.analyze$make_analyze_steps$fn__52750.invoke(analyze.clj:106)
	at metabase.sync.util$fn__42218$run_step_with_metadata__42223$fn__42227$fn__42229.invoke(util.clj:345)
	at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:102)
	at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:96)
	at metabase.sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:119)
	at metabase.sync.util$with_start_and_finish_debug_logging.invoke(util.clj:116)
	at metabase.sync.util$fn__42218$run_step_with_metadata__42223$fn__42227.invoke(util.clj:342)
	at metabase.sync.util$fn__42218$run_step_with_metadata__42223.invoke(util.clj:337)
	at metabase.sync.util$fn__42410$run_sync_operation__42415$fn__42416$fn__42417.invoke(util.clj:430)
	at clojure.core$mapv$fn__8445.invoke(core.clj:6912)
	at clojure.lang.PersistentVector.reduce(PersistentVector.java:343)
	at clojure.core$reduce.invokeStatic(core.clj:6827)
	at clojure.core$mapv.invokeStatic(core.clj:6903)
	at clojure.core$mapv.invoke(core.clj:6903)
	at metabase.sync.util$fn__42410$run_sync_operation__42415$fn__42416.invoke(util.clj:430)
	at metabase.sync.util$fn__42410$run_sync_operation__42415.invoke(util.clj:424)
	at metabase.sync.analyze$fn__52762$analyze_db_BANG___52767$fn__52768$fn__52769.invoke(analyze.clj:123)
	at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:149)
	at metabase.sync.util$do_with_error_handling.invoke(util.clj:142)
	at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:145)
	at metabase.sync.util$do_with_error_handling.invoke(util.clj:142)
	at clojure.core$partial$fn__5839.invoke(core.clj:2623)
	at metabase.driver$fn__22445.invokeStatic(driver.clj:540)
	at metabase.driver$fn__22445.invoke(driver.clj:540)
	at clojure.lang.MultiFn.invoke(MultiFn.java:239)
	at metabase.sync.util$sync_in_context$fn__42130.invoke(util.clj:138)
	at metabase.sync.util$with_db_logging_disabled$fn__42127.invoke(util.clj:129)
	at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:102)
	at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:96)
	at metabase.sync.util$with_start_and_finish_logging$fn__42116.invoke(util.clj:114)
	at metabase.sync.util$with_sync_events$fn__42111.invoke(util.clj:88)
	at metabase.sync.util$with_duplicate_ops_prevented$fn__42102.invoke(util.clj:67)
	at metabase.sync.util$do_sync_operation.invokeStatic(util.clj:164)
	at metabase.sync.util$do_sync_operation.invoke(util.clj:161)
	at metabase.sync.analyze$fn__52762$analyze_db_BANG___52767$fn__52768.invoke(analyze.clj:120)
	at metabase.sync.analyze$fn__52762$analyze_db_BANG___52767.invoke(analyze.clj:115)
	at metabase.api.database$fn__55211$fn__55213.invoke(database.clj:615)
	at clojure.core$binding_conveyor_fn$fn__5754.invoke(core.clj:2030)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.sql.SQLException: Cannot read LocalDate using a Types.SMALLINT field
	at org.mariadb.jdbc.internal.com.read.resultset.rowprotocol.TextRowProtocol.getInternalLocalDate(TextRowProtocol.java:1143)
	at org.mariadb.jdbc.internal.com.read.resultset.SelectResultSet.getObject(SelectResultSet.java:1283)
	at com.mchange.v2.c3p0.impl.NewProxyResultSet.getObject(NewProxyResultSet.java:217)
	at metabase.driver.sql_jdbc.execute$get_object_of_class_thunk$fn__72525.invoke(execute.clj:280)
	at clojure.core$juxt$fn__5824.invoke(core.clj:2600)
	at metabase.driver.sql_jdbc.execute$row_thunk$row_thunk_STAR___72583.invoke(execute.clj:348)
	at metabase.query_processor.reducible$reducible_rows$reify__38029.reduce(reducible.clj:156)
	at clojure.core$transduce.invokeStatic(core.clj:6883)
	at clojure.core$transduce.invokeStatic(core.clj:6879)
	at clojure.core$transduce.invoke(core.clj:6870)
	at metabase.query_processor.context.default$default_reducef$fn__37927.invoke(default.clj:59)
	... 137 more
```

</details>

<details><summary>Query stacktrace</summary>

```
09-03 16:03:11 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 18,
 :started_at #t "2020-09-03T16:03:10.883479+02:00[Europe/Copenhagen]",
 :via
 [{:status :failed,
   :class clojure.lang.ExceptionInfo,
   :error "Error reducing result rows",
   :stacktrace
   ["--> query_processor.context.default$default_reducef$fn__37927.invoke(default.clj:61)"
    "query_processor.context.default$default_reducef.invokeStatic(default.clj:58)"
    "query_processor.context.default$default_reducef.invoke(default.clj:49)"
    "query_processor.context$reducef.invokeStatic(context.clj:69)"
    "query_processor.context$reducef.invoke(context.clj:62)"
    "query_processor.context.default$default_runf$respond_STAR___37931.invoke(default.clj:70)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:392)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)"
    "driver.sql_jdbc$fn__73834.invokeStatic(sql_jdbc.clj:49)"
    "driver.sql_jdbc$fn__73834.invoke(sql_jdbc.clj:47)"
    "query_processor.context$executef.invokeStatic(context.clj:59)"
    "query_processor.context$executef.invoke(context.clj:48)"
    "query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
    "query_processor.context.default$default_runf.invoke(default.clj:67)"
    "query_processor.context$runf.invokeStatic(context.clj:45)"
    "query_processor.context$runf.invoke(context.clj:39)"
    "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
    "query_processor.reducible$pivot.invoke(reducible.clj:31)"
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__45532.invoke(mbql_to_native.clj:26)"
    "query_processor.middleware.check_features$check_features$fn__44834.invoke(check_features.clj:42)"
    "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__45697.invoke(optimize_datetime_filters.clj:133)"
    "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47225.invoke(wrap_value_literals.clj:137)"
    "query_processor.middleware.annotate$add_column_info$fn__43455.invoke(annotate.clj:574)"
    "query_processor.middleware.permissions$check_query_permissions$fn__44709.invoke(permissions.clj:64)"
    "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__46215.invoke(pre_alias_aggregations.clj:40)"
    "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__44907.invoke(cumulative_aggregations.clj:61)"
    "query_processor.middleware.resolve_joins$resolve_joins$fn__46747.invoke(resolve_joins.clj:183)"
    "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39185.invoke(add_implicit_joins.clj:245)"
    "query_processor.middleware.limit$limit$fn__45518.invoke(limit.clj:38)"
    "query_processor.middleware.format_rows$format_rows$fn__45499.invoke(format_rows.clj:81)"
    "query_processor.middleware.desugar$desugar$fn__44973.invoke(desugar.clj:22)"
    "query_processor.middleware.binning$update_binning_strategy$fn__43999.invoke(binning.clj:229)"
    "query_processor.middleware.resolve_fields$resolve_fields$fn__44515.invoke(resolve_fields.clj:24)"
    "query_processor.middleware.add_dimension_projections$add_remapping$fn__38734.invoke(add_dimension_projections.clj:272)"
    "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__38941.invoke(add_implicit_clauses.clj:141)"
    "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39334.invoke(add_source_metadata.clj:105)"
    "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__46412.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
    "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__43640.invoke(auto_bucket_datetimes.clj:125)"
    "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44562.invoke(resolve_source_table.clj:46)"
    "query_processor.middleware.parameters$substitute_parameters$fn__46197.invoke(parameters.clj:114)"
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__44614.invoke(resolve_referenced.clj:80)"
    "query_processor.middleware.expand_macros$expand_macros$fn__45229.invoke(expand_macros.clj:158)"
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__39365.invoke(add_timezone_info.clj:15)"
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__47109.invoke(splice_params_in_response.clj:32)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46423$fn__46427.invoke(resolve_database_and_driver.clj:33)"
    "driver$do_with_driver.invokeStatic(driver.clj:61)"
    "driver$do_with_driver.invoke(driver.clj:57)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46423.invoke(resolve_database_and_driver.clj:27)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45447.invoke(fetch_source_query.clj:267)"
    "query_processor.middleware.store$initialize_store$fn__47118$fn__47119.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
    "query_processor.store$do_with_store.invoke(store.clj:40)"
    "query_processor.middleware.store$initialize_store$fn__47118.invoke(store.clj:10)"
    "query_processor.middleware.cache$maybe_return_cached_results$fn__44491.invoke(cache.clj:209)"
    "query_processor.middleware.validate$validate_query$fn__47127.invoke(validate.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__45545.invoke(normalize_query.clj:22)"
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39203.invoke(add_rows_truncated.clj:36)"
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__47094.invoke(results_metadata.clj:147)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__44850.invoke(constraints.clj:42)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__46286.invoke(process_userland_query.clj:136)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__44793.invoke(catch_exceptions.clj:174)"
    "query_processor.reducible$async_qp$qp_STAR___38013$thunk__38014.invoke(reducible.clj:101)"
    "query_processor.reducible$async_qp$qp_STAR___38013.invoke(reducible.clj:107)"
    "query_processor.reducible$sync_qp$qp_STAR___38022$fn__38025.invoke(reducible.clj:133)"
    "query_processor.reducible$sync_qp$qp_STAR___38022.invoke(reducible.clj:132)"
    "query_processor$process_userland_query.invokeStatic(query_processor.clj:213)"
    "query_processor$process_userland_query.doInvoke(query_processor.clj:209)"
    "query_processor$fn__47269$process_query_and_save_execution_BANG___47278$fn__47281.invoke(query_processor.clj:225)"
    "query_processor$fn__47269$process_query_and_save_execution_BANG___47278.invoke(query_processor.clj:217)"
    "query_processor$fn__47313$process_query_and_save_with_max_results_constraints_BANG___47322$fn__47325.invoke(query_processor.clj:237)"
    "query_processor$fn__47313$process_query_and_save_with_max_results_constraints_BANG___47322.invoke(query_processor.clj:230)"
    "api.dataset$fn__50604$fn__50607.invoke(dataset.clj:54)"
    "query_processor.streaming$streaming_response_STAR_$fn__35435$fn__35436.invoke(streaming.clj:73)"
    "query_processor.streaming$streaming_response_STAR_$fn__35435.invoke(streaming.clj:72)"
    "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:66)"
    "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:64)"
    "async.streaming_response$do_f_async$fn__23249.invoke(streaming_response.clj:85)"],
   :error_type :qp,
   :ex-data {:type :qp}}],
 :state nil,
 :error_type :qp,
 :json_query {:database 18, :query {:source-table 125}, :type "query", :parameters [], :middleware {:add-default-userland-constraints? true}},
 :native
 {:query
  "SELECT `x_formatting`.`created_at` AS `created_at`, `x_formatting`.`deleted_at` AS `deleted_at`, `x_formatting`.`ID` AS `ID`, `x_formatting`.`order_date` AS `order_date`, `x_formatting`.`order_time` AS `order_time`, `x_formatting`.`order_time_micro` AS `order_time_micro`, `x_formatting`.`price` AS `price`, `x_formatting`.`total` AS `total`, `x_formatting`.`updated_at` AS `updated_at`, `x_formatting`.`order_year` AS `order_year`, `x_formatting`.`time_int` AS `time_int`, `x_formatting`.`time_bigint` AS `time_bigint` FROM `x_formatting` LIMIT 2000",
  :params nil},
 :status :failed,
 :class java.sql.SQLException,
 :stacktrace
 ["org.mariadb.jdbc.internal.com.read.resultset.rowprotocol.TextRowProtocol.getInternalLocalDate(TextRowProtocol.java:1143)"
  "org.mariadb.jdbc.internal.com.read.resultset.SelectResultSet.getObject(SelectResultSet.java:1283)"
  "com.mchange.v2.c3p0.impl.NewProxyResultSet.getObject(NewProxyResultSet.java:217)"
  "--> driver.sql_jdbc.execute$get_object_of_class_thunk$fn__72525.invoke(execute.clj:280)"
  "driver.sql_jdbc.execute$row_thunk$row_thunk_STAR___72583.invoke(execute.clj:348)"
  "query_processor.reducible$reducible_rows$reify__38029.reduce(reducible.clj:156)"
  "query_processor.context.default$default_reducef$fn__37927.invoke(default.clj:59)"
  "query_processor.context.default$default_reducef.invokeStatic(default.clj:58)"
  "query_processor.context.default$default_reducef.invoke(default.clj:49)"
  "query_processor.context$reducef.invokeStatic(context.clj:69)"
  "query_processor.context$reducef.invoke(context.clj:62)"
  "query_processor.context.default$default_runf$respond_STAR___37931.invoke(default.clj:70)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:392)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)"
  "driver.sql_jdbc$fn__73834.invokeStatic(sql_jdbc.clj:49)"
  "driver.sql_jdbc$fn__73834.invoke(sql_jdbc.clj:47)"
  "query_processor.context$executef.invokeStatic(context.clj:59)"
  "query_processor.context$executef.invoke(context.clj:48)"
  "query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
  "query_processor.context.default$default_runf.invoke(default.clj:67)"
  "query_processor.context$runf.invokeStatic(context.clj:45)"
  "query_processor.context$runf.invoke(context.clj:39)"
  "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
  "query_processor.reducible$pivot.invoke(reducible.clj:31)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__45532.invoke(mbql_to_native.clj:26)"
  "query_processor.middleware.check_features$check_features$fn__44834.invoke(check_features.clj:42)"
  "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__45697.invoke(optimize_datetime_filters.clj:133)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47225.invoke(wrap_value_literals.clj:137)"
  "query_processor.middleware.annotate$add_column_info$fn__43455.invoke(annotate.clj:574)"
  "query_processor.middleware.permissions$check_query_permissions$fn__44709.invoke(permissions.clj:64)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__46215.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__44907.invoke(cumulative_aggregations.clj:61)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__46747.invoke(resolve_joins.clj:183)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39185.invoke(add_implicit_joins.clj:245)"
  "query_processor.middleware.limit$limit$fn__45518.invoke(limit.clj:38)"
  "query_processor.middleware.format_rows$format_rows$fn__45499.invoke(format_rows.clj:81)"
  "query_processor.middleware.desugar$desugar$fn__44973.invoke(desugar.clj:22)"
  "query_processor.middleware.binning$update_binning_strategy$fn__43999.invoke(binning.clj:229)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__44515.invoke(resolve_fields.clj:24)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__38734.invoke(add_dimension_projections.clj:272)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__38941.invoke(add_implicit_clauses.clj:141)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39334.invoke(add_source_metadata.clj:105)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__46412.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__43640.invoke(auto_bucket_datetimes.clj:125)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44562.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.parameters$substitute_parameters$fn__46197.invoke(parameters.clj:114)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__44614.invoke(resolve_referenced.clj:80)"
  "query_processor.middleware.expand_macros$expand_macros$fn__45229.invoke(expand_macros.clj:158)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__39365.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__47109.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46423$fn__46427.invoke(resolve_database_and_driver.clj:33)"
  "driver$do_with_driver.invokeStatic(driver.clj:61)"
  "driver$do_with_driver.invoke(driver.clj:57)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46423.invoke(resolve_database_and_driver.clj:27)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45447.invoke(fetch_source_query.clj:267)"
  "query_processor.middleware.store$initialize_store$fn__47118$fn__47119.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__47118.invoke(store.clj:10)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__44491.invoke(cache.clj:209)"
  "query_processor.middleware.validate$validate_query$fn__47127.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__45545.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39203.invoke(add_rows_truncated.clj:36)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__47094.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__44850.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__46286.invoke(process_userland_query.clj:136)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__44793.invoke(catch_exceptions.clj:174)"
  "query_processor.reducible$async_qp$qp_STAR___38013$thunk__38014.invoke(reducible.clj:101)"
  "query_processor.reducible$async_qp$qp_STAR___38013.invoke(reducible.clj:107)"
  "query_processor.reducible$sync_qp$qp_STAR___38022$fn__38025.invoke(reducible.clj:133)"
  "query_processor.reducible$sync_qp$qp_STAR___38022.invoke(reducible.clj:132)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:213)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:209)"
  "query_processor$fn__47269$process_query_and_save_execution_BANG___47278$fn__47281.invoke(query_processor.clj:225)"
  "query_processor$fn__47269$process_query_and_save_execution_BANG___47278.invoke(query_processor.clj:217)"
  "query_processor$fn__47313$process_query_and_save_with_max_results_constraints_BANG___47322$fn__47325.invoke(query_processor.clj:237)"
  "query_processor$fn__47313$process_query_and_save_with_max_results_constraints_BANG___47322.invoke(query_processor.clj:230)"
  "api.dataset$fn__50604$fn__50607.invoke(dataset.clj:54)"
  "query_processor.streaming$streaming_response_STAR_$fn__35435$fn__35436.invoke(streaming.clj:73)"
  "query_processor.streaming$streaming_response_STAR_$fn__35435.invoke(streaming.clj:72)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:66)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:64)"
  "async.streaming_response$do_f_async$fn__23249.invoke(streaming_response.clj:85)"],
 :context :ad-hoc,
 :error "Cannot read LocalDate using a Types.SMALLINT field",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:database 18,
  :query
  {:source-table 125,
   :fields
   [[:datetime-field [:field-id 823] :default]
    [:datetime-field [:field-id 824] :default]
    [:field-id 820]
    [:datetime-field [:field-id 857] :default]
    [:datetime-field [:field-id 859] :default]
    [:datetime-field [:field-id 858] :default]
    [:field-id 821]
    [:field-id 825]
    [:datetime-field [:field-id 822] :default]
    [:field-id 19760]
    [:field-id 19762]
    [:field-id 19761]],
   :limit 2000},
  :type :query,
  :middleware {:add-default-userland-constraints? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :nested? false,
   :query-hash [-20, -7, -9, 44, 116, 115, -43, 109, -103, -93, -124, 111, -34, -83, 90, -56, 93, -51, -122, 29, 21, -37, 96, 19, -96, -56, 55, -105, 78, -78, 13, -38]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :data {:rows [], :cols []}}
```

</details>

Possible workaround is to either change for column type from `year` to `int(4)` - or to it with a view.

**Information about your Metabase Installation:**
Metabase 0.36.4 with MySQL5.7 and MariaDB10.4

**Additional context**
https://discourse.metabase.com/t/error-there-was-a-problem-with-your-question-cannot-read-localdate-using-a-types-smallint-field/11451

--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -368,6 +368,24 @@
         (catch Throwable _
           (.getString rs i))))))
 
+(defmethod sql-jdbc.execute/read-column-thunk [:mysql Types/DATE]
+  [driver ^ResultSet rs rsmeta ^Integer i]
+  (let [parent-thunk ((get-method sql-jdbc.execute/read-column-thunk [:sql-jdbc Types/DATE]) driver rs rsmeta i)]
+    (fn read-time-thunk []
+      ;; YEAR type is stored as smallint and exposed in the driver as either java.sql.Date or a short, depending on
+      ;; the mysql configuration option `yearIsDateType`. The parent thunk will try (.getObject rs i
+      ;; java.time.LocalDate) but this errors with "Cannot read LocalDate using a Types.SMALLINT
+      ;; field". https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-type-conversions.html. (NOTE: the
+      ;; documentation mentions the return type is java.sql.Short but I believe this is a typo for java.lang.Short:
+      ;; https://bugs.mysql.com/bug.php?id=35115)
+      (if (= "YEAR" (.getColumnTypeName rsmeta i))
+        (let [x (.getObject rs i)]
+          (condp instance? x
+            java.sql.Date (.toLocalDate x)
+            java.lang.Short (long x)
+            :else x))
+        (parent-thunk)))))
+
 (defn- format-offset [t]
   (let [offset (t/format "ZZZZZ" (t/zone-offset t))]
     (if (= offset "Z")

--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -372,9 +372,8 @@
   [driver ^ResultSet rs ^ResultSetMetaData rsmeta ^Integer i]
   (if (= "YEAR" (.getColumnTypeName rsmeta i))
     (fn read-time-thunk []
-      (let [x (.getObject rs i)]
-        (when x
-          (.toLocalDate ^java.sql.Date x))))
+      (when-let [x (.getObject rs i)]
+        (.toLocalDate ^java.sql.Date x)))
     (let [parent-thunk ((get-method sql-jdbc.execute/read-column-thunk [:sql-jdbc Types/DATE]) driver rs rsmeta i)]
       (parent-thunk))))
 

--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -371,20 +371,20 @@
 (defmethod sql-jdbc.execute/read-column-thunk [:mysql Types/DATE]
   [driver ^ResultSet rs ^ResultSetMetaData rsmeta ^Integer i]
   (let [parent-thunk ((get-method sql-jdbc.execute/read-column-thunk [:sql-jdbc Types/DATE]) driver rs rsmeta i)]
-    (fn read-time-thunk []
-      ;; YEAR type is stored as smallint and exposed in the driver as either java.sql.Date or a short, depending on
-      ;; the mysql configuration option `yearIsDateType`. The parent thunk will try (.getObject rs i
-      ;; java.time.LocalDate) but this errors with "Cannot read LocalDate using a Types.SMALLINT
-      ;; field". https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-type-conversions.html. (NOTE: the
-      ;; documentation mentions the return type is java.sql.Short but I believe this is a typo for java.lang.Short:
-      ;; https://bugs.mysql.com/bug.php?id=35115)
-      (if (= "YEAR" (.getColumnTypeName rsmeta i))
+    (if (= "YEAR" (.getColumnTypeName rsmeta i))
+      (fn read-time-thunk []
+        ;; YEAR type is stored as smallint and exposed in the driver as either java.sql.Date or a short, depending on
+        ;; the mysql configuration option `yearIsDateType`. The parent thunk will try (.getObject rs i
+        ;; java.time.LocalDate) but this errors with "Cannot read LocalDate using a Types.SMALLINT
+        ;; field". https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-type-conversions.html. (NOTE: the
+        ;; documentation mentions the return type is java.sql.Short but I believe this is a typo for java.lang.Short:
+        ;; https://bugs.mysql.com/bug.php?id=35115)
         (let [x (.getObject rs i)]
           (condp instance? x
             java.sql.Date (.toLocalDate ^java.sql.Date x)
             java.lang.Short (long x)
-            :else x))
-        (parent-thunk)))))
+            :else x)))
+      (parent-thunk))))
 
 (defn- format-offset [t]
   (let [offset (t/format "ZZZZZ" (t/zone-offset t))]

--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -259,7 +259,7 @@
     :TINYTEXT   :type/Text
     :VARBINARY  :type/*
     :VARCHAR    :type/Text
-    :YEAR       :type/Integer}
+    :YEAR       :type/Date}
    ;; strip off " UNSIGNED" from end if present
    (keyword (str/replace (name database-type) #"\sUNSIGNED$" ""))))
 
@@ -370,20 +370,12 @@
 
 (defmethod sql-jdbc.execute/read-column-thunk [:mysql Types/DATE]
   [driver ^ResultSet rs ^ResultSetMetaData rsmeta ^Integer i]
-  (let [parent-thunk ((get-method sql-jdbc.execute/read-column-thunk [:sql-jdbc Types/DATE]) driver rs rsmeta i)]
-    (if (= "YEAR" (.getColumnTypeName rsmeta i))
-      (fn read-time-thunk []
-        ;; YEAR type is stored as smallint and exposed in the driver as either java.sql.Date or a short, depending on
-        ;; the mysql configuration option `yearIsDateType`. The parent thunk will try (.getObject rs i
-        ;; java.time.LocalDate) but this errors with "Cannot read LocalDate using a Types.SMALLINT
-        ;; field". https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-type-conversions.html. (NOTE: the
-        ;; documentation mentions the return type is java.sql.Short but I believe this is a typo for java.lang.Short:
-        ;; https://bugs.mysql.com/bug.php?id=35115)
-        (let [x (.getObject rs i)]
-          (when x
-            (condp instance? x
-              java.sql.Date (.toLocalDate ^java.sql.Date x)
-              java.lang.Short (long x)))))
+  (if (= "YEAR" (.getColumnTypeName rsmeta i))
+    (fn read-time-thunk []
+      (let [x (.getObject rs i)]
+        (when x
+          (.toLocalDate ^java.sql.Date x))))
+    (let [parent-thunk ((get-method sql-jdbc.execute/read-column-thunk [:sql-jdbc Types/DATE]) driver rs rsmeta i)]
       (parent-thunk))))
 
 (defn- format-offset [t]

--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -369,7 +369,7 @@
           (.getString rs i))))))
 
 (defmethod sql-jdbc.execute/read-column-thunk [:mysql Types/DATE]
-  [driver ^ResultSet rs rsmeta ^Integer i]
+  [driver ^ResultSet rs ^ResultSetMetaData rsmeta ^Integer i]
   (let [parent-thunk ((get-method sql-jdbc.execute/read-column-thunk [:sql-jdbc Types/DATE]) driver rs rsmeta i)]
     (fn read-time-thunk []
       ;; YEAR type is stored as smallint and exposed in the driver as either java.sql.Date or a short, depending on
@@ -381,7 +381,7 @@
       (if (= "YEAR" (.getColumnTypeName rsmeta i))
         (let [x (.getObject rs i)]
           (condp instance? x
-            java.sql.Date (.toLocalDate x)
+            java.sql.Date (.toLocalDate ^java.sql.Date x)
             java.lang.Short (long x)
             :else x))
         (parent-thunk)))))

--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -380,10 +380,10 @@
         ;; documentation mentions the return type is java.sql.Short but I believe this is a typo for java.lang.Short:
         ;; https://bugs.mysql.com/bug.php?id=35115)
         (let [x (.getObject rs i)]
-          (condp instance? x
-            java.sql.Date (.toLocalDate ^java.sql.Date x)
-            java.lang.Short (long x)
-            :else x)))
+          (when x
+            (condp instance? x
+              java.sql.Date (.toLocalDate ^java.sql.Date x)
+              java.lang.Short (long x)))))
       (parent-thunk))))
 
 (defn- format-offset [t]

--- a/test/metabase/db/metadata_queries_test.clj
+++ b/test/metabase/db/metadata_queries_test.clj
@@ -1,10 +1,17 @@
 (ns metabase.db.metadata-queries-test
   (:require [clojure.test :refer :all]
             [metabase
-             [models :refer [Field Table]]
+             [driver :as driver]
+             [models :as models :refer [Field Table]]
+             [query-processor :as qp]
              [test :as mt]]
             [metabase.db.metadata-queries :as metadata-queries]
-            [metabase.driver.sql-jdbc.test-util :as sql-jdbc.tu]))
+            [metabase.driver.sql-jdbc.test-util :as sql-jdbc.tu]
+            [metabase.models
+             [database :as database]
+             [field :as field]
+             [table :as table]]
+            [schema.core :as s]))
 
 ;; Redshift tests are randomly failing -- see https://github.com/metabase/metabase/issues/2767
 (defn- metadata-queries-test-drivers []
@@ -39,12 +46,12 @@
                   ["33 Taps"]
                   ["800 Degrees Neapolitan Pizzeria"]
                   ["BCD Tofu House"]]
-        table (Table (mt/id :venues))
-        fields [(Field (mt/id :venues :name))]
-        fetch! #(->> (metadata-queries/table-rows-sample table fields (when % {:truncation-size %}))
-                     ;; since order is not guaranteed do some sorting here so we always get the same results
-                     (sort-by first)
-                     (take 5))]
+        table    (Table (mt/id :venues))
+        fields   [(Field (mt/id :venues :name))]
+        fetch!   #(->> (metadata-queries/table-rows-sample table fields (when % {:truncation-size %}))
+                       ;; since order is not guaranteed do some sorting here so we always get the same results
+                       (sort-by first)
+                       (take 5))]
     (is (= :type/Text (-> fields first :base_type)))
     (mt/test-drivers (sql-jdbc.tu/sql-jdbc-drivers)
       (is (= expected (fetch! nil)))
@@ -53,4 +60,20 @@
           (is (= (mapv (fn [[s]] [(subs (or s "") 0 (min size (count s)))])
                        expected)
                  (fetch! size))
-              "Did not truncate a text field"))))))
+              "Did not truncate a text field")))))
+  (testing "substring checking"
+    ;; turn off validation so we can just return the query that is emitted and check for expressions
+    (s/without-fn-validation
+      ;; just return the query map
+      (with-redefs [qp/process-query (fn [query] {:data {:rows (:query query)}})
+                    table/database   (constantly (database/map->DatabaseInstance {:id 5678}))]
+        (let [table  (table/map->TableInstance {:id 1234})
+              fields [(field/map->FieldInstance {:id 4321 :base_type :type/Text})]]
+          (testing "uses substrings if driver supports expressions"
+            (with-redefs [driver/supports? (constantly true)]
+              (let [query (metadata-queries/table-rows-sample table fields {:truncation-size 4})]
+                (is (seq (:expressions query))))))
+          (testing "doesnt' use substrings if driver doesn't support expressions"
+            (with-redefs [driver/supports? (constantly false)]
+              (let [query (metadata-queries/table-rows-sample table fields {:truncation-size 4})]
+                (is (empty (:expressions query)))))))))))

--- a/test/metabase/driver/mysql_test.clj
+++ b/test/metabase/driver/mysql_test.clj
@@ -9,11 +9,13 @@
              [sync :as sync]
              [test :as mt]
              [util :as u]]
+            [metabase.db.metadata-queries :as metadata-queries]
             [metabase.driver.sql-jdbc.connection :as sql-jdbc.conn]
             [metabase.models
              [database :refer [Database]]
              [field :refer [Field]]
              [table :refer [Table]]]
+            [metabase.sync.analyze.fingerprint :as fingerprint]
             [metabase.test.data.interface :as tx]
             [toucan
              [db :as db]
@@ -81,6 +83,27 @@
                    {:name "id", :base_type :type/Integer, :special_type :type/PK}
                    {:name "thing", :base_type :type/Text, :special_type :type/Category}}
                  (db->fields db))))))))
+
+(tx/defdataset ^:private year-db
+  [["years"
+    [{:field-name "year_column", :base-type {:native "YEAR"}}]
+    [[2001] [2002] [1999]]]])
+
+(deftest year-test
+  (mt/test-driver :mysql
+    (mt/dataset year-db
+      (testing "By default YEAR"
+        (is (= #{{:name "year_column", :base_type :type/Integer, :special_type :type/Category}
+                 {:name "id", :base_type :type/Integer, :special_type :type/PK}}
+               (db->fields (mt/db)))))
+      (let [table  (db/select-one Table :db_id (u/id (mt/db)))
+            fields (db/select Field :table_id (u/id table))]
+        (testing "Can select from this table"
+          (is (= [[#t "2001-01-01" 1] [#t "2002-01-01" 2] [#t "1999-01-01" 3]]
+                 (metadata-queries/table-rows-sample table fields))))
+        (testing "We can fingerprint this table"
+          (is (= 2
+                 (:updated-fingerprints (#'fingerprint/fingerprint-table! table fields)))))))))
 
 (deftest db-timezone-id-test
   (mt/test-driver :mysql

--- a/test/metabase/driver/mysql_test.clj
+++ b/test/metabase/driver/mysql_test.clj
@@ -93,16 +93,16 @@
   (mt/test-driver :mysql
     (mt/dataset year-db
       (testing "By default YEAR"
-        (is (= #{{:name "year_column", :base_type :type/Integer, :special_type :type/Category}
+        (is (= #{{:name "year_column", :base_type :type/Date, :special_type :type/Category}
                  {:name "id", :base_type :type/Integer, :special_type :type/PK}}
                (db->fields (mt/db)))))
       (let [table  (db/select-one Table :db_id (u/id (mt/db)))
-            fields (db/select Field :table_id (u/id table))]
+            fields (db/select Field :table_id (u/id table) :name "year_column")]
         (testing "Can select from this table"
-          (is (= [[#t "2001-01-01" 1] [#t "2002-01-01" 2] [#t "1999-01-01" 3]]
+          (is (= [[#t "2001-01-01"] [#t "2002-01-01"] [#t "1999-01-01"]]
                  (metadata-queries/table-rows-sample table fields))))
         (testing "We can fingerprint this table"
-          (is (= 2
+          (is (= 1
                  (:updated-fingerprints (#'fingerprint/fingerprint-table! table fields)))))))))
 
 (deftest db-timezone-id-test

